### DISCUSSION TITLE
How to work with form data?

### DISCUSSION THREAD

**Discussion by zaitompro** [fastapi/fastapi#8286] [Questions] [answered]
Hello everyone!
I need to process form data but I can't do that with fastapi. To try understand how can I do it I used Postman to send data in different ways:
from form-data tab
from x-www-form-urlencoded
with header "Content-Type: application/x-www-form-urlencoded"
with header "Content-Type: application/json"
without headers

In any case I have the same result: {"detail": "There was an error parsing the body"}.

![postman](https://user-images.githubusercontent.com/22984590/57228831-54404d80-7047-11e9-8fa6-4502b8a29dca.png)

```
from fastapi import FastAPI, HTTPException, Form

@app.post("/subscription_email")
def subscription_email(*, customers: int = Form(...)):
    return customers
```
Code based on example from docs.

Please, help me to fix it and understand how to work woth form data from POST requests.

**Comment by wshayes** [comment]
Here is what I'm using for file upload (using forms):

https://fastapi.tiangolo.com/tutorial/request-forms-and-files/

    @router.post("/nanopubs/import/file", tags=["Nanopubs"])
    def import_nanopubs(
        file: UploadFile = File(...),
        overwrite: bool = Query(False, description="Overwrite if exists when true"),
        validation: bool = Query(True, description="Validate nanopub when true"),
        email: str = Query("", description="Will send email with processing results"),
    ):

        if file:
            contents = file.file.read().decode("utf-8")
            try:
                nanopubs = json.loads(contents)
            except Exception as e:
                log.error("Error reading nanopub json content", error=str(e))

**Comment by zaitompro** [comment]
@wshayes thank you for advice. It works fine with query parameters but still doesn't work with data from form fields.

**Comment by euri10** [comment]
your example works fine out of the box with the swagger generated at /docs

edit: email doesnt support markdown, see screens below

**Comment by euri10** [comment]
![Imgur](https://i.imgur.com/oncN00s.png)

same with postman

![Imgur](https://i.imgur.com/cNDCG6G.png)

**Comment by zaitompro** [comment]
@euri10 thank you that you try to help me. I tried to make request trough swagger and it doesen't work too. Strange situation. GET-requests work fine, POST-requests with JSON work fine (with pydantic BaseModal). But, it doesen't work for me with form data.

**Comment by euri10** [comment]
I copy pasted your code and that works fine as shown on the screens so
really don't understand what's happening.
Have you logs, errors?

**Comment by zaitompro** [comment]
@euri10 can you tell me please where is the logs of fastapi? Even ```find / -name "*.log"``` command doesn't show anything looks like fastapi log.
I don't know is it important or not but I use dockerized fastapi based on official tiangolo docker image https://hub.docker.com/r/tiangolo/uvicorn-gunicorn-fastapi. We use docker containers in the company for CI/CD.

**Comment by euri10** [comment]
if you're using that image just do a `docker logs`
> But, it doesen't work for me with form data.

it's rather hard to help you without knowing what doesn't work exactly, a better description of your issue might help :grin:

**Comment by zaitompro** [comment, marked as answer]
@euri10 thank you very much!!! Now it works!!!
I checked logs and found reason.
The reason is "python-multipart" was not installed. I was surprised because I thought that image contains any requirements for fastapi.
Anyway, now I know how to check what's wrong:) Thank you so much!!!

**Comment by zaitompro** [comment]
@euri10
How to get nested values from form data?
We have requests like this.
![nested form data](https://user-images.githubusercontent.com/22984590/57363469-bddd6a80-71b3-11e9-8482-76fd4fb73892.png)

If I do something like this (based on docs example)
```
from fastapi import FastAPI, HTTPException, Form
from typing import Dict, List

app = FastAPI()

@app.post("/subscription_email")
def subscription_email(*, subdomain: Dict[str,str] = Form(...)):
    return subdomain
```
It doesn't work. I have status 422 and response:
```
{
    "detail": [
        {
            "loc": [
                "body",
                "subdomain"
            ],
            "msg": "field required",
            "type": "value_error.missing"
        }
    ]
}
```
May I get whole data in one dict by one step?
Unfortunatelly, we can't change request in that way to expect JSON because it's a webhook from thirdparty service.

**Comment by tiangolo** [comment]
Thanks @wshayes and @euri10 for your help!

@zaitompro there's no standard way to encode nested values in form data. It's not a limitation of FastAPI but of the web standards (HTTP). In your case, it is probably being encoded somehow: comma separated values, space-separated values, JSON strings embedded in form fields, etc.

What you can do is declare them as `str`, and then extract those strings and parse them with whatever is the specific way those fields are encoded.

To start, you can just declare them as `str`, log them. With that, you can infer how those values are actually encoded.

**Comment by zaitompro** [comment]
@tiangolo thank you for advice!
How can I get whole body without validation? I didn't find it in documentation. I tried few options what I found docs but all of that with validation and expectation.
I switched to Python from PHP and on PHP I had no problems with nested form data. Maybe it's the reason why I can't understand why it's a problem in Python that uses the same HTTP.
Please, give me little example or link with example how to get whole body or get body parameters without validation and expectation.

**Comment by tiangolo** [comment]
@zaitompro you can use the request directly: https://fastapi.tiangolo.com/tutorial/using-request-directly/

And from it, you can extract the body without validation.

**Comment by zaitompro** [comment]
@tiangolo thank you so much for help!!!
I got all data that I need and now I know how to get any data from body.

**Comment by tiangolo** [comment]
Great! I'm glad to hear that. Thanks for reporting back and closing the issue.

**Comment by Getitdan** [comment]
> @zaitompro you can use the request directly: https://fastapi.tiangolo.com/tutorial/using-request-directly/
> 
> And from it, you can extract the body without validation.

The link is broken, the new one is: https://fastapi.tiangolo.com/advanced/using-request-directly/

**Comment by solankydev** [comment]
> @euri10 thank you very much!!! Now it works!!!
> I checked logs and found reason.
> The reason is "python-multipart" was not installed. I was surprised because I thought that image contains any requirements for fastapi.
> Anyway, now I know how to check what's wrong:) Thank you so much!!!

I was also having same issue and then realized that I need to install "python-multipart". Thanks.

**Comment by yandongliu** [comment]
Ran into the same error. Quite dumb to not report the missing package.

**Comment by mecampbellsoup** [comment]
> @zaitompro you can use the request directly: [fastapi.tiangolo.com/tutorial/using-request-directly](https://fastapi.tiangolo.com/tutorial/using-request-directly/)
> 
> And from it, you can extract the body without validation.

This is required because I cannot declare a form parameter with a nested name (because it's not valid Python syntax), correct?

```
@app.post('/subscriptions', response_class=HTMLResponse)
async def create_subscription(
        request: Request,
		# NOT VALID PYTHON
        'subscription[customer_attributes][reference]': Form(...)
        ):
```
My form data names are nested, e.g.:

```
[gatekeeper] (b'product_handle=coreweave-cloud&subscription%5Bcustomer_attributes%5D%5Brefer'
[gatekeeper]  b'ence%5D=tenant-test&subscription%5Bcustomer_attributes%5D%5Bfirst_name%5D=Ma'
[gatekeeper]  b'tt&subscription%5Bcustomer_attributes%5D%5Blast_name%5D=Campbell&subscriptio'
[gatekeeper]  b'n%5Bcustomer_attributes%5D%5Bemail%5D=&subscription%5Bcustomer_attributes%5D'
[gatekeeper]  b'%5Bphone%5D=&subscription%5Bcustomer_attributes%5D%5Borganization%5D=&subscr'
[gatekeeper]  b'iption%5Bpayment_profile_attributes%5D%5Bfirst_name%5D=&subscription%5Bpayme'
[gatekeeper]  b'nt_profile_attributes%5D%5Blast_name%5D=&subscription%5Bpayment_profile_attr'
[gatekeeper]  b'ibutes%5D%5Bfull_number%5D=&subscription%5Bpayment_profile_attributes%5D%5Bc'
[gatekeeper]  b'vv%5D=&subscription%5Bpayment_profile_attributes%5D%5Bexpiration_month%5D=&s'
[gatekeeper]  b'ubscription%5Bpayment_profile_attributes%5D%5Bexpiration_year%5D=&subscripti'
[gatekeeper]  b'on%5Bpayment_profile_attributes%5D%5Bbilling_zip%5D=&subscription%5Bagree_to'
[gatekeeper]  b'_terms%5D=on')
```
Assuming I am unable to rename my form input names, I _must_ fallback to using raw request body/data, right?

**Comment by Yakobro-coder** [comment]
> > @zaitompro you can use the request directly: [fastapi.tiangolo.com/tutorial/using-request-directly](https://fastapi.tiangolo.com/tutorial/using-request-directly/)
> > And from it, you can extract the body without validation.
> 
> This is required because I cannot declare a form parameter with a nested name (because it's not valid Python syntax), correct?
> 
> ```
> @app.post('/subscriptions', response_class=HTMLResponse)
> async def create_subscription(
>         request: Request,
> 		# NOT VALID PYTHON
>         'subscription[customer_attributes][reference]': Form(...)
>         ):
> ```
> 
> My form data names are nested, e.g.:
> 
> ```
> [gatekeeper] (b'product_handle=coreweave-cloud&subscription%5Bcustomer_attributes%5D%5Brefer'
> [gatekeeper]  b'ence%5D=tenant-test&subscription%5Bcustomer_attributes%5D%5Bfirst_name%5D=Ma'
> [gatekeeper]  b'tt&subscription%5Bcustomer_attributes%5D%5Blast_name%5D=Campbell&subscriptio'
> [gatekeeper]  b'n%5Bcustomer_attributes%5D%5Bemail%5D=&subscription%5Bcustomer_attributes%5D'
> [gatekeeper]  b'%5Bphone%5D=&subscription%5Bcustomer_attributes%5D%5Borganization%5D=&subscr'
> [gatekeeper]  b'iption%5Bpayment_profile_attributes%5D%5Bfirst_name%5D=&subscription%5Bpayme'
> [gatekeeper]  b'nt_profile_attributes%5D%5Blast_name%5D=&subscription%5Bpayment_profile_attr'
> [gatekeeper]  b'ibutes%5D%5Bfull_number%5D=&subscription%5Bpayment_profile_attributes%5D%5Bc'
> [gatekeeper]  b'vv%5D=&subscription%5Bpayment_profile_attributes%5D%5Bexpiration_month%5D=&s'
> [gatekeeper]  b'ubscription%5Bpayment_profile_attributes%5D%5Bexpiration_year%5D=&subscripti'
> [gatekeeper]  b'on%5Bpayment_profile_attributes%5D%5Bbilling_zip%5D=&subscription%5Bagree_to'
> [gatekeeper]  b'_terms%5D=on')
> ```
> 
> Assuming I am unable to rename my form input names, I _must_ fallback to using raw request body/data, right?


I managed to solve the problem as follows.
I hope this helps someone.
```
from fastapi import FastAPI, Form
import uvicorn

app = FastAPI()


@app.post("/test")
async def leads_data(responsible_id: str = Form(alias='leads[responsible][0][id]'),
                responsible_user_id: str = Form(alias='leads[responsible][0][responsible_user_id]'),
                account_id: str = Form(alias='account[id]'),
                subdomain: str = Form(alias='account[subdomain]')):
    result = dict(responsible_id=responsible_id, responsible_user_id=responsible_user_id,
                  account_id=account_id, subdomain=subdomain)

    print(result)
    print()
    return {"STATUS": "OK"}
```
